packages/task/src/browser/task-service.ts

```diff
@@ -727,34 +730,48 @@ export class TaskService implements TaskConfigurationClient {
     }
 
     async runTask(task: TaskConfiguration, option?: RunTaskOption): Promise<TaskInfo | undefined> {
-        const runningTasksInfo: TaskInfo[] = await this.getRunningTasks();
+        console.debug('entering runTask');
+        const lockToken: number = await this.taskStartingLock.acquire();
```

Comment:
Wall of code incoming. For locks like these, we usually use a different but equivalent pattern in this case.

Since the lock cannot be configured and only chains tasks one after the other, we can use promise chaining:

```ts
    // move this line to the top of the class with the other fields
    protected taskExecutionQueue: Promise<void> = Promise.resolve();
    async runTask(task: TaskConfiguration, option?: RunTaskOption): Promise<TaskInfo | undefined> {
        // backup the current task execution queue
        const currenTaskExecutionQueue = this.taskExecutionQueue;
        // create the lock representing this task
        const lock = new Deferred<void>();
        // update the queue to chain ourselves
        this.taskExecutionQueue = currenTaskExecutionQueue.then(() => lock.promise);
        // wait for our turn to execute
        await currenTaskExecutionQueue;
        try {
            const runningTasksInfo: TaskInfo[] = await this.getRunningTasks();

            // check if the task is active
            const matchedRunningTaskInfo = runningTasksInfo.find(taskInfo => {
                const taskConfig = taskInfo.config;
                return this.taskDefinitionRegistry.compareTasks(taskConfig, task);
            });
            if (matchedRunningTaskInfo) { // the task is active
                lock.resolve();
                const taskName = this.taskNameResolver.resolve(task);
                const terminalId = matchedRunningTaskInfo.terminalId;
                if (terminalId) {
                    const terminal = this.terminalService.getByTerminalId(terminalId);
                    if (terminal) {
                        if (TaskOutputPresentation.shouldSetFocusToTerminal(task)) { // assign focus to the terminal if presentation.focus is true
                            this.terminalService.open(terminal, { mode: 'activate' });
                        } else if (TaskOutputPresentation.shouldAlwaysRevealTerminal(task)) { // show the terminal but not assign focus
                            this.terminalService.open(terminal, { mode: 'reveal' });
                        }
                    }
                }
                const selectedAction = await this.messageService.info(`The task '${taskName}' is already active`, 'Terminate Task', 'Restart Task');
                if (selectedAction === 'Terminate Task') {
                    await this.terminateTask(matchedRunningTaskInfo);
                } else if (selectedAction === 'Restart Task') {
                    return this.restartTask(matchedRunningTaskInfo, option);
                }
            } else { // run task as the task is not active
                return this.doRunTask(task, option);
            }
        } catch (error) {
            console.error(error);
        } finally {
            lock.resolve();
        }
    }
```

So far we avoided creating such a construct, but if you really want to introduce a new `Lock` class then I might have to look closer at the current implementation.

Reply:
I prefer my approach because it replaces a pattern with a piece of code that can be reused. Is there a reason we use that pattern instead a lock class?

Reply:
I guess in this specific scenario, the promise chain pattern is not the most adequate since we need a mechanism to release early. Agreed that a dedicated class sounds good here. See https://github.com/eclipse-theia/theia/pull/9858#issuecomment-897840612.